File: leetcode/src/main/java/nonlinear/tree/invertbinarytree226/Solution3.java
package nonlinear.tree.invertbinarytree226;

import nonlinear.tree.invertbinarytree226.Main.TreeNode;

public class Solution3 {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    public TreeNode invertTree(TreeNode root) {
        // 빈 노드가 아닐때 처리
        if (root != null) {
            invertTree(root.left);  // 왼쪽 자식 노드 DFS
            invertTree(root.right); // 오른쪽 자식 노드 DFS

            // 왼쪽/오른쪽 자식 노드 스왑
            TreeNode temp = root.left;
            root.left = root.right;
            root.right = temp;
        }
        return root;
    }
}
// 0ms

File: leetcode/src/main/java/nonlinear/tree/constructbinarytreefrompreorderandinordertraversal105/Solution3.kt
package nonlinear.tree.constructbinarytreefrompreorderandinordertraversal105

class Solution3 {


    fun buildTree(preorder: IntArray, inorder: IntArray): Main.TreeNode? {
        // 예외 처리
        if (inorder.isEmpty()) return null

        // 전위 순회 값이 중위 순회에서는 몇 번째 인덱스인지 추출
        val inIndex = inorder.indexOf(preorder[0])
        // 해당 인덱스는 중위 순회를 분할하는 노드로 지정
        val node = Main.TreeNode(inorder[inIndex])

        // 왼쪽 자식 노드부터 진행
        node.left = buildTree(preorder.copyOfRange(1, inIndex + 1), inorder.copyOfRange(0, inIndex))
        // 오른쪽 자식 노드 진행
        node.right = buildTree(preorder.copyOfRange(inIndex + 1, preorder.size), inorder.copyOfRange(inIndex + 1, inorder.size))
        return node
    }
}